lib/dirac_twisted_clover.cpp

```diff
@@ -234,32 +292,27 @@ namespace quda {
     int odd_bit = (matpcType == QUDA_MATPC_ODD_ODD || matpcType == QUDA_MATPC_ODD_ODD_ASYMMETRIC) ? 1 : 0;
     QudaParity parity[2] = {static_cast<QudaParity>((1 + odd_bit) % 2), static_cast<QudaParity>((0 + odd_bit) % 2)};
 
-    FullClover *cs = new FullClover(clover, false);
-    FullClover *cI = new FullClover(clover, true);
-
     if (in.TwistFlavor() == QUDA_TWIST_SINGLET) {
-      if (symmetric) {
-	if (dagger) {
-	  TwistCloverInv(*tmp1, in, parity[1]);
-	  Dslash(out, *tmp1, parity[0]);
-	  TwistCloverInv(*tmp1, out, parity[0]);
-	  DslashXpay(out, *tmp1, parity[1], in, kappa2);
-	} else {
-	  Dslash(*tmp1, in, parity[0]);
-	  DslashXpay(out, *tmp1, parity[1], in, kappa2);
-	}
-      } else { // asymmetric preconditioning
-        double a = 2.0 * kappa * in.TwistFlavor() * mu;
+      if (!symmetric) { // asymmetric preconditioning
 	Dslash(*tmp1, in, parity[0]);
-	twistedCloverDslashCuda(&static_cast<cudaColorSpinorField&>(out), *gauge, cs, cI,
-				static_cast<cudaColorSpinorField*>(tmp1), parity[1], dagger,
-				&static_cast<const cudaColorSpinorField&>(in),
-				QUDA_DEG_DSLASH_CLOVER_TWIST_XPAY, a, kappa2, 0.0, 0.0, commDim, profile);
-
-	flops += (1320ll+96ll)*in.Volume();
+        DiracTwistedClover::DslashXpay(out, *tmp1, parity[1], in, kappa2);
+      } else if (!dagger) { // symmetric preconditioning
+        Dslash(*tmp1, in, parity[0]);
+        DslashXpay(out, *tmp1, parity[1], in, kappa2);
+      } else { // symmetric preconditioning, dagger
+#ifdef NEW_DSLASH
+        TwistCloverInv(out, in, parity[1]);
+        reverse = true; Dslash(*tmp1, out, parity[0]); reverse = false;
+        DiracWilson::DslashXpay(out, *tmp1, parity[1], in, kappa2);
+#else
+        TwistCloverInv(*tmp1, in, parity[1]);
+        DiracWilson::Dslash(out, *tmp1, parity[0]);
+        TwistCloverInv(*tmp1, out, parity[0]);
+        DiracWilson::DslashXpay(out, *tmp1, parity[1], in, kappa2);
+#endif
       }
     } else { //Twist doublet
-      errorQuda("Non-degenerate DiracTwistedCloverPC is not implemented \n");
+      errorQuda("Non-degenerate operator is not implemented");
```

Comment:
Can this error be more instructive? i.e., mentioning that the user should explicitly use the non-degenerate operator instead of the regular twisted operator? This comment is relevant other places for twisted mass, too.

Reply:
Well this non-degenerate twisted clover is something that  is going to fill in at some point, and noone is using  it at the moment, so I don't see this as something to really worry about.

Reply:
Sounds good!